File: combination-sum.js
// Time Complexity : O(2^N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
let result;
var combinationSum = function(candidates, target) {
    // Edge Case
    if (!candidates || !candidates.length || target == 0) return [[]];
    result = [];
    backtrack(candidates, target, 0, 0, []);
    return result;
};

var backtrack = function(candidates, target, sum, index, temp) {
    // Base Case
    if (sum == target) {
        result.push([...temp]);
        return;
    }
    if (sum > target) return;
    
    // Logic
    for (let i = index; i < candidates.length; i++) {
        // Action
        temp.push(candidates[i]);
        // Recurse
        backtrack(candidates, target, sum + candidates[i], i, temp);
        // Backtrack
        temp.pop();
    }
};

File: CombinationSum.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    /// Time Complexity : O(n)
    // Space Complexity :O(n)
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  No
    public class CombinationSum
    {
        IList<IList<int>> result;
        public IList<IList<int>> CombinationSumREC(int[] candidates, int target)
        {
            result = new List<IList<int>>();
            if (candidates == null || candidates.Length == 0) return result;
            helper(candidates, target, 0, new List<int>());
            return result;
        }

        private void helper(int[] candidates, int target, int idx, List<int> path)
        {
            //base
            if (target == 0)
            {
                result.Add(new List<int>(path));
                return;
            }
            if (target < 0) return;
            //logic
            for (int i = idx; i < candidates.Length; i++)
            {
                //action
                path.Add(candidates[i]);
                //recurse
                helper(candidates, target - candidates[i], i, path);
                //backtrack
                path.RemoveAt(path.Count - 1);
            }
        }
    }
}
